### ISSUE TITLE
RTCPeerConnection

### ISSUE THREAD

**Issue by abrahamjuliot** [abrahamjuliot/creepjs#67]
Concept:
```js
const getWebRTCData = () => {
    const rtcPeerConnection = window.RTCPeerConnection || window.webkitRTCPeerConnection || window.mozRTCPeerConnection || window.msRTCPeerConnection
    const connection = new rtcPeerConnection({
        iceServers: [{
            urls: ["stun:stun.l.google.com:19302?transport=udp"]
        }]
    }, {
        optional: [{
            RtpDataChannels: !0
        }]
    })
    connection.onicecandidate = e => {
        const candidateEncoding = /((udp|tcp)\s)((\d|\w)+\s)((\d|\w|(\.|\:))+)(?=\s)/ig
		const connectionLineEncoding = /(c=IN\s)(.+)\s/ig

        if (!e.candidate) {
            return
        }
        const {
            candidate,
            address: ipAddress
        } = e.candidate

        const encodingMatch = candidate.match(candidateEncoding)
        if (encodingMatch) {
			const { sdp } = e.target.localDescription
			
            const candidateIpAddress = encodingMatch[0].split(' ')[2]
			const connectionLineIpAddress = sdp.match(connectionLineEncoding)[0].trim().split(' ')[2]
			
            console.log(ipAddress)
            console.log(candidateIpAddress)
			console.log(connectionLineIpAddress)
        }
    }
    connection.createDataChannel('bl')
    connection.createOffer().then(e => connection.setLocalDescription(e))
}
getWebRTCData()


```

**Comment by abrahamjuliot** [comment]
try in iframe

**Comment by abrahamjuliot** [comment]
What other data points can add to the fingerprint?

**Comment by Thorin-Oakenpants** [comment]
Oh goodie, you're doing webRTC stuff, which has been on my list for ages
- https://thorin-oakenpants.github.io/testing/ <-- super shitty initial webRTC support test (from like a year ago)
- some bookmarks in my ToDo/WebRTC
   - https://test.webrtc.org/
   - https://webrtc.github.io/samples/

**Comment by Thorin-Oakenpants** [comment]
oh .. and also this is one of my stars on github - https://github.com/muaz-khan/DetectRTC

**Comment by abrahamjuliot** [comment]
If I make the promise async and await the offer, I can capture client ip address in Firefox via the `connection` and `candidate`.

**Comment by Niek** [comment]
FYI (not sure if this is worth opening a new issue): when WebRTC is disabled in Chrome I get this error: `1: Error - RTCIceCandidate failed`. In Firefox (with WebRTC disabled) it works without error.

**Comment by abrahamjuliot** [comment]
Nice catch. It appears Firefox disables the `RTCPeerConnection` interface whereas Chrome disables the `RTCIceCandidate` connection. I refactored the `onicecandidate` event handler function to resolve if `event.candidate` is null. This should go live by next month, but you may test it now at https://creepjs--abrahamjuliot.repl.co (unstable development site). Let me know if the error persists on your end.

**Comment by Niek** [comment]
> Nice catch. It appears Firefox disables the `RTCPeerConnection` interface whereas Chrome disables the `RTCIceCandidate` connection. I refactored the `onicecandidate` event handler function to resolve if `event.candidate` is null. This should go live by next month, but you may test it now at https://creepjs--abrahamjuliot.repl.co (unstable development site). Let me know if the error persists on your end.

Thanks, this is indeed fixed now. I have more false positives on Firefox on Android:
* `userAgent contains gibberish: [fx] Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) FxQuantum/86.0 AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.0.0 Mobile Safari/537.36`
* `Expected "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) FxQuantum/86.0 AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.0.0 Mobile Safari/537.36" in nested iframe and got "Mozilla/5.0 (Android 10; Mobile; rv:86.0) Gecko/86.0 Firefox/86.0"`

Weirdly enough this seems to happen kinda random, most of the times it gets the "regular" UA, other times it gets the "FxQuantum" one.

**Comment by abrahamjuliot** [comment]
I've noticed a few similar cases as well: Firefox plugins in Pale Moon, VirtualBox WebGL renderer, and I imagine there are many more. To reduce this, I made the gibberish test less strict in 3bebee6 (#66). The letter sequence gibbers will be ignored if less than 3 exist. 

I've also considered removing the gibberish test since it is a loose metric and does not provide stable entropy, but I've kept it to highlight gibbers consistently caught in the plugins and WebGL vendor/renderer in Brave. I am open to removing though if it throws too many false positives in other contexts.
